# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose
0,1,-8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875.639453,покупка жилья
1,1,-4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080.014102,приобретение автомобиля
2,0,-5623.422610,33,Среднее,1,женат / замужем,0,M,сотрудник,0,145885.952297,покупка жилья
3,3,-4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628.550329,дополнительное образование
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616.077870,сыграть свадьбу
5,0,-926.185831,27,высшее,0,гражданский брак,1,M,компаньон,0,255763.565419,покупка жилья
6,0,-2879.202052,43,высшее,0,женат / замужем,0,F,компаньон,0,240525.971920,операции с жильем
7,0,-152.779569,50,СРЕДНЕЕ,1,женат / замужем,0,M,сотрудник,0,135823.934197,образование
8,2,-6929.865299,35,ВЫСШЕЕ,0,гражданский брак,1,F,сотрудник,0,95856.832424,на проведение свадьбы
9,0,-2188.756445,41,среднее,1,женат / замужем,0,M,сотрудник,0,144425.938277,покупка жилья для семьи


**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

### Удаление пропусков (продолжение)

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для того, чтобы упорядочитьть категории по возрастанию риска невозврата по кредиту, необходимо сгруппировать клиентов по искомым категориям, посчитать их просрочки и найти процент клиентов от текущей группы, которые имеют задолженность.

In [22]:
def rate(row,param):
    data = row.groupby(param).agg({'debt':'sum'})
    data['count'] = row.groupby(param)['debt'].count()
    data['rate'] = data['debt'] / data['count'] * 100
    return data

In [23]:
data_children = rate(data,'children')
data_children.sort_values(by='rate',ascending=True)

,debt,count,rate
children,,,
5,0,9,0.000000
0,1063,14091,7.543822
3,27,330,8.181818
1,444,4808,9.234609
2,194,2052,9.454191
4,4,41,9.756098


**Вывод:** 

Клиенты без детей выплачивают кредиты более добросовестно, чем с детьми. Возможно, это связано с увеличенными рисками внезапных трат.

 <div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

 Здорово, что пишешь функции для автоматизации задачи. Вывод верный, но обращу внимание, что группы с 3-5 детьми составляют не более 2 процентов от общей выворки, поэтому их можно не учитывать в выводе.
    
Можно также сделать расчёты с помощью нескольких функций по одному или нескольким столбцам.
    
Примеры:
    
* data.groupby('column1')['column2'].agg(['count', 'mean', 'sum'])
    
* data.groupby('column1').agg({'column1' : 'count', 'column2' : 'mean',})
    
* data.pivot_table(index = 'column1' , values = 'column2', aggfunc=['count', 'mean', 'sum'])
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
data_family = rate(data,'family_status')
data_family.sort_values(by='rate',ascending=True)

,debt,count,rate
family_status,,,
вдовец / вдова,63,951,6.624606
в разводе,84,1189,7.064760
женат / замужем,927,12261,7.560558
гражданский брак,385,4134,9.313014
Не женат / не замужем,273,2796,9.763948


**Вывод:** 

Наименьшие показатели по задолженностям в группе вдов и вдовцов, в то время как у клиентов, которые не находятся в отношениях приблизительно в 1.5 раза выше, чем у вдов/вдовцов. Вероятно, это связано с тем, что наличие текущих серьезных отношений или отношений в прошлом заставляют людей быть более дисциплинированными.

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

 С выводом согласен, тебе верно удалось определить главный признак, по которому можно разделить все группы на 2 категории. 👍
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
data_income = rate(data,'total_income_category')
data_income.sort_values(by='rate',ascending=True)

,debt,count,rate
total_income_category,,,
D,21,349,6.017192
B,354,5014,7.060231
A,2,25,8.000000
C,1353,15921,8.498210
E,2,22,9.090909


В данных есть группы заемщиков, в которых менее 100 человек. Эти данные можно использовать для анализа, но они не релевантны, на мой взгляд. Убираем эти группы и готовим таблицу для передачи в анализ и скоринг.

In [26]:
data_income.loc[data_income['count'] > 100].sort_values(by='rate',ascending=True)

,debt,count,rate
total_income_category,,,
D,21,349,6.017192
B,354,5014,7.060231
C,1353,15921,8.498210


**Вывод:** 

Наиболее надежными клиентами оказываются люди с уровнем дохода от 30 до 50 тр.

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Верно. Правильно, что обращаешь внимание на объёмы категорий клиентов. Так можно быть уверенным в объективности выводов.
    
Так как группы несбалансированны, можно решить это задание через разделение выборки на равные по объёму части. Для этого подойдёт функция pd.qcut().
    
Здесь можно об этом почитать:
    
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html

Метод qcut() делит числовой столбец на равные по объёму части на основании квантилей. Квантили — это статистические показатели, которые разделяют набор данных на равные части. Квантили включают квартили (25%, 50% и 75%), децили (каждые 10%) и процентили (каждые 1%). Приведу пример разделения на 3 равных отрезка, находящихся на отметках 33,3% и 66,7%.
        
    pd.qcut(data['column'], q=3, labels=['one', 'two', 'three'])
        
Можно не указывать названия отрезков с помощью labels. Тогда названия отрезкам присвоятся в численном виде.
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
data_purpose = rate(data,'purpose_category')
data_purpose.sort_values(by='rate',ascending=True)

,debt,count,rate
purpose_category,,,
операции с недвижимостью,780,10751,7.255139
проведение свадьбы,183,2313,7.911803
получение образования,369,3988,9.252758
операции с автомобилем,400,4279,9.347978


**Вывод:** 

Кредиты по операциям с недвижимостью составляют 50% от общего количества и при этом клиенты ответственнее подходят к выплатам. Вероятно, в большинстве случаев, в связи с крупной суммой займа, банк требует залог от клиента, что дисциплинирует и мотивирует. Займы по автокредитам имеют наибольшую долю задолженностей.

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Вывод верный. Здорово, что стараешься описывать причины, которые могли привести к полученной картине. Для аналитика важно видеть смысл, который стоит за данными.
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Ошибки в данных возникают по технологическим причинам и из-за человеческого фактора.
Технологические ошибки возникают при копировании, считывании и смене формата данных.

Причина пропуска может быть технологической - это сбой работы БД при выгрузке, сбой работы приложений, при внесении данных в базу, неудачная миграция данных в процессе переноса данных или обновления АБС.

Человеческий фактор приводит к ошибкам, таким как опечатки, неправильные числа или категориальные значения. Например, оператор не внёс размер дохода и стаж заёмщика, или определённый контрагент не собирал эти данные.Возможно, пропуски связаны с определенной категорией (например, заемщики по образовательным кредитам, которые не имеют стажа и дохода)


<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Абсолютно верно. Основные причины пропусков - это технические проблемы и человеческий фактор.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства.
Медиана более объективна. Медиана  —  это такое число в выборке, что ровно половина элементов больше него, а другая половина —  меньше.
Медианное значение не искажает картину распределения по пользователям. Например, если в выборке появится несколько людей с экстремально высокими или низкими показателями, они не повлияют (или не значительно повлияют) на медианное значение, когда среднеарифметическое может резко вырасти или уменьшиться.

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Абсолютно верно, такие сильно выделяющиеся значения ещё называют выбросами.
</div>

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

##### Для улучшения системы скоринга клиентов была проанализирована выборка текущих клиентов и их задолженностей по платежам.

##### В рамках анализа были проверены следующие гипотезы:

###### Есть ли зависимость между количеством детей и возвратом кредита в срок?

Категория клиентов без детей оказалась самой многочисленной и процент задержек по кредиту самый низкий - 7.54%. Самый высокий процент задержек по оплатам оказался у клиентов с двумя детьми и составляет 9.45%.

###### Есть ли зависимость между семейным положением и возвратом кредита в срок?

По результатам проверки данной гипотезы выявлена следующая закономерность - меньше всего просрочек по кредитам у вдов/вдовцев (6.62%), больше всего у клиентов, не состоящих в отношениях (9.76%). Самой массовой категорией являются клиенты в браке, процент просрочек по данной категории составляет 7.56%

###### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

По результатам проверки данной гипотезы выявлена следующая закономерность - меньше всего просрочек по кредитам у клиентов с уровнем дохода от 30 000 до 50 000 руб в месяц (6.02%), больше всего у клиентов с уровнем дохода от 200 000 до 1 000 000 руб в месяц (8.50%).

Категория клиентов с экстремальными значениями доходов (ниже 30 000 руб или выше 1 000 000 руб) являются слишком малой. Достоверно говорить о вероятностях просрочек от этих клиентов нельзя.

###### Как разные цели кредита влияют на его возврат в срок?

По результатам проверки данной гипотезы выявлена следующая закономерность - меньше всего просрочек по кредитам по займам на недвижимость (7.26%), больше всего по займам на автомобиль (9.35%).

Есть зависимости между целями кредита, семейным положением, уровнем дохода и количеством детей в семье. Нельзя сказать, что они кардинально отличаются, но 2% на больших числах могут оказать существенное влияние на кредитный портфель банка. Особое внимание следует уделить вдовам/вдовцам или молодоженам — с ними могут риски по невозврату кредитов выше.

##### Рекомендации для банка

- Уделить внимание стандартам валидации на сервере, чтобы в будущем не сталкиваться с проблемой дубликатов.

- Продумать сценарий, когда отличается дата/время заявки, но уникальный ID пользователя тот же, иначе фактически дубликаты будут, но метод dublicated() их не найдет.

- Считать стаж работы в годах, пользователи чаще всего не знают свой стаж в днях.

- Показывать операторам в банке или пользователям на сайте подсказки заранее, когда вводят сверхвысокие/сверхнизкие
  значения в полях: возраст, стаж, количество детей.
  
- Блокировать отправку заявок с заведомо некорректными значениями, которые могут быть важны для анализа, например,
  с отрицательным количеством детей.

<div class="alert alert-warning">
    <h3> Комментарий ревьюера ⚠️<a class="tocSkip"> </h3>

Хороший структурированный вывод, учитывающий все вопросы исследовательской части.
        
Для его улучшения стоит также добавить информацию о том, что было сделано на этапе предобработки, так как это позволит упростить взаимодействие с потенциальным заказчиком. В следующий раз датасет мог бы прийти более чистым, а тебе можно было бы приступить сразу к аналитической части работы.
        
И ещё можно дать рекомендации по портретам хорошего и плохого клиентов.

Если захочешь поработать над оформлением, советую изучить [краткое руководство по Маркдауну](https://paulradzkov.com/2014/markdown_cheatsheet/).
</div>